Add support for boot ISOs to AnacondaWebUIWorkflow

velezd opened the pull request.

velezd committed:
Move BootIsoStructure from kickstart_test to separate plugin

velezd committed:
Add support for boot ISOs to AnacondaWebUIWorkflow

Currently boot ISOs are supported only when local hypersvisor is used.

jstodola commented on libpermian/plugins/anaconda_webui/__init__.py:
```diff
@@ -391,20 +401,46 @@ def displayStatus(self):
+
+        if self.boot_iso_structure[self.architecture].startswith(('http://', 'https://')):
+            self.boot_iso_path = os.path.join(self.temp_dir, f'boot-{self.architecture}.iso')
+            if not os.path.exists(self.boot_iso_path):
```
It might be useful to log an info message when boot.iso already exists and won't be downloaded again.

velezd replied:
I've added log message

jstodola commented on libpermian/plugins/anaconda_webui/__init__.py:
```diff
@@ -391,20 +401,46 @@ def displayStatus(self):
         return self.crc.result.state == 'canceled'
 
+    def _get_boot_iso_path(self):
+        """ Gets path to boot iso for virt-install command """
```
This is a bit confusing to me, since the method is called `_get_....`, but doesn't return anything.

velezd replied:
Renamed to `_set_...`

jstodola reviewed (dismissed):
It looks good to me, thanks.

rvykydal commented on libpermian/plugins/anaconda_webui/__init__.py:
```diff
@@ -391,20 +401,56 @@ def displayStatus(self):
+            kernel_cmdline += f' inst.stage2={os_url}'
         else:
-            location = os_url
+            location = self.boot_iso_path
```
I am afraid we may need to set path to kernel and initrd also for boot.iso. There has been a change of location in the iso images quite recently and it seems libvirt is not able to find the kernel on default place on some of our runners (F34). Maybe we could use InstallationStructure to supply the path via its kernel and initrd settings but I am not sure if it is clever, confusing, or dirty soulution.

Other option would be to require runners >= F37. I guess default location should work on F37 and newer.

velezd replied:
I don't like the idea of adding workarounds for EOL versions of Fedora. The runners should be upgraded. Or if that would be a problem maybe we could limit what runners are used for this workflow?

rvykydal commented:
@velezd do you think we could copy the local iso to the temporary workspace of the workflow, like in https://github.com/rvykydal/permian/commit/9e4cff85a8dc28977baaf2419b707ea0080aa048. It would solve this issue: https://github.com/rhinstaller/anaconda/pull/4834#issuecomment-1594662017

velezd commented:
> @velezd do you think we could copy the local iso to the temporary workspace of the workflow, like in [rvykydal@9e4cff8](https://github.com/rvykydal/permian/commit/9e4cff85a8dc28977baaf2419b707ea0080aa048). It would solve this issue: [rhinstaller/anaconda#4834 (comment)](https://github.com/rhinstaller/anaconda/pull/4834#issuecomment-1594662017)

Yes, I think this is the best solution. Done.

rvykydal commented:
So testing the workflow on F37, I've hit another issue: https://github.com/rhinstaller/anaconda/pull/4834#issuecomment-1597222794

velezd committed:
Add function for checking virt-install log for errors + restarting

And implement restart loop for _start_vm function. This will prevent
issues when multiple VMs are being started at the same time and also
make the test fail faster in case of real errors.

velezd committed:
Move boot-scratch race-condition handling into _check_vm_errors

rvykydal commented:
> So testing the workflow on F37, I've hit another issue: [rhinstaller/anaconda#4834 (comment)](https://github.com/rhinstaller/anaconda/pull/4834#issuecomment-1597222794)

Seems to be fixed: https://github.com/rhinstaller/anaconda/pull/4834#issuecomment-1598671910

rvykydal reviewed (dismissed):
Looks good to me, thank you.
